subcommand/inject-connect/command.go

```diff
@@ -363,6 +398,98 @@ func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, client
 	}
 }
 
+func (c *Command) parseAndValidateResourceFlags() (corev1.ResourceRequirements, corev1.ResourceRequirements, error) {
+	// Init container
+	var initContainerCPULimit, initContainerCPURequest, initContainerMemoryLimit, initContainerMemoryRequest resource.Quantity
```

Comment:
There is a lot of code repetition between this function and the code in `Run` that validates default sidecar resources. Another refactoring idea is to make another function that will parse and validate resources for a single "entity" and then call it for each of the init container, lifecycle sidecar, and potentially for the sidecar proxy as well.

It could look something like this:
```go
func validateAndParseResource(flagPrefix, cpuRequest, cpuLimit, memoryRequest, memoryLimit string) (corev1.ResourceRequirements, error) {
	var containerCPULimit, containerCPURequest, containerMemoryLimit, containerMemoryRequest resource.Quantity

	// Parse and validate the initContainer resources.
	containerCPURequest, err := resource.ParseQuantity(cpuRequest)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("%s-cpu-request is invalid: %s", flagPrefix, err)
	}
	containerCPULimit, err = resource.ParseQuantity(cpuLimit)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("%s-cpu-limit is invalid: %s", flagPrefix, err)
	}
	if containerCPULimit.Value() != 0 && containerCPURequest.Cmp(containerCPULimit) > 0 {
		return corev1.ResourceRequirements{}, fmt.Errorf(
			"request must be <= limit: %s-cpu-request value of %q is greater than the %s-cpu-limit value of %q",
			flagPrefix, cpuRequest, flagPrefix, cpuLimit)
	}

	containerMemoryRequest, err = resource.ParseQuantity(memoryRequest)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("%s-memory-request is invalid: %s", flagPrefix, err)
	}
	containerMemoryLimit, err = resource.ParseQuantity(memoryLimit)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("%s-memory-limit is invalid: %s", flagPrefix, err)
	}
	if containerMemoryLimit.Value() != 0 && containerMemoryRequest.Cmp(containerMemoryLimit) > 0 {
		return corev1.ResourceRequirements{}, fmt.Errorf(
			"request must be <= limit: %s-memory-request value of %q is greater than the %s-memory-limit value of %q",
			flagPrefix, memoryRequest, flagPrefix, memoryLimit)
	}

	// Put into corev1.ResourceRequirements form
	return corev1.ResourceRequirements{
		Requests: corev1.ResourceList{
			corev1.ResourceCPU:    containerCPURequest,
			corev1.ResourceMemory: containerMemoryRequest,
		},
		Limits: corev1.ResourceList{
			corev1.ResourceCPU:    containerCPULimit,
			corev1.ResourceMemory: containerMemoryLimit,
		},
	}, nil
}
```
Then in this function, you'd just call:
```go
initResources, err := validateAndParseResource("-init-container", c.flagInitContainerCPURequest, c.flagInitContainerCPULimit, c.flagInitContainerMemoryRequest, c.flagInitContainerMemoryLimit)
```

Reply:
This could potentially be improved even more because there's also duplication of validations between CPU and memory resources that could probably be moved into another function as well.

Reply:
There are two repeats in this function. The "don't repeat yourself" rule of thumb generally is suggested to start at 3 since it's over-optimization before then. This suggestion also decrease the readability of the code and makes it more fragile.

Reply:
Per my comment above, there is also sidecar resource validation which is almost exactly the same and it lives in the `Run` function, which makes it 3 repeats.

Why are you saying it'll make the code less readable and fragile? I found it pretty hard to read as it's written right now which why I suggested a refactor.

Reply:
I understand that there is some logic sharing. There are also subtle differences in behavior between the new flags and the previous proxy flags, specifically in the defaulting behavior. I think it's a reasonable suggestion to refactor this but I don't think it should be part of this PR. If you would like me to log an issue for this, let me know and I will do so.

Reply:
After looking at this some more, I'd like to see this refactor happen.

To preserve the ability of not setting the sidecar proxy resources we could add an additional `allowEmpty bool` argument to the suggested `validateAndParseResource` function. This could be used like:

```go
	if !allowEmpty || cpuRequest != "" {
		// Parse and validate the initContainer resources.
		containerCPURequest, err = resource.ParseQuantity(cpuRequest)
		if err != nil {
			return corev1.ResourceRequirements{}, fmt.Errorf("%s-cpu-request is invalid: %s", flagPrefix, err)
		}
	}
```

Alternatively we could default those flags to "0". This isn't a breaking change because there's no difference between 0 and not being set. There's a long comment somewhere why I did it like this but in the interest of less complexity and conformity I don't think it's necessary to keep that if it makes the implementation too complex.